docs/spec/constructors.rst

```diff
@@ -337,7 +337,7 @@ Consistency of ``__new__`` and ``__init__``
 -------------------------------------------
 
 Type checkers may optionally validate that the ``__new__`` and ``__init__``
-methods for a class have consistent signatures.
+methods for a class have :term:`consistent` signatures.
```

Comment:
I initially thought the same, and then after thinking more I decided it is correct in the formal sense? I think these two methods should have "the same" signature, and "consistent" is the formal term for "the same, accounting for gradual types."

Unless there is instead a unidirectional assignability relationship between the signatures of these methods, and one can be wider than the other? If so, which direction does the assignability relation go?

Reply:
I checked this in pyright, mypy, and pyre, with the following code sample:

```
from typing import Any, Self

arg: Any

class C:
    def __new__(cls, x: str) -> Self: return cls(arg)
    def __init__(self, x: str) -> None: ...

class D:
    def __new__(cls, x: str = "foo") -> Self: return cls(arg)
    def __init__(self, x: str) -> None: ...

class E:
    def __new__(cls, x: str) -> Self: return cls(arg)
    def __init__(self, x: str = "foo") -> None: ...
    
class F:
    def __new__(cls, x: str) -> Self: return cls(arg)
    def __init__(self, x: int) -> None: ...

class G:
    def __new__(cls, x: str) -> Self: return cls(arg)
    def __init__(self, x: Any) -> None: ...

class H:
    def __new__(cls, x: Any) -> Self: return cls(arg)
    def __init__(self, x: str) -> None: ...
```

Under default settings, none of the three appear to validate `__init__` and `__new__` signatures against each other at all -- all raise no errors on this example.

Under strict mode, pyright errors on all three of D, E, and F, with an error on the `__init__` method that says "Mismatch between signature of __new__ and __init__".

Given that `D.__new__` is assignable to `D.__init__` and `E.__init__` is assignable to E.__new__`, but pyright errors on both (and does not error on `G` and `H`, whose signatures differ but are consistent) I conclude that "consistent with" is indeed the right formal term for the check pyright implements here.

And since pyright seems to be the only one that implements this check, it seems like this is reasonable wording for the spec.

It also seems reasonable for the runtime behavior. You could implement a class where one of `__init__` or `__new__` is a wider type than the other, but in this case you would practically be limited to using the narrower signature for both.

Reply:
Yes, pyright currently checks that the two methods are "assignable to each other". The specific code is [here](https://github.com/microsoft/pyright/blob/ecc4ebb12a28b5bd65ee9399f4f8d8246892fc35/packages/pyright-internal/src/analyzer/checker.ts#L5597) if you're interested.

As you noted, the `reportInconsistentConstructor` check is not on by default in pyright — it's a strict-mode check. So while I agree that you could implement a class where `__init__` or `__new__` is wider than the other, that's probably not a good idea, and it's reasonable for a strict-mode check to flag it as a problem. At least that's my thinking.